core/src/main/java/io/grpc/ChannelImpl.java

```diff
@@ -293,6 +293,7 @@ public void start(Listener<RespT> observer, Metadata.Headers headers) {
         return;
       }
       try {
+        completeHeaders(headers);
```

Comment:
Could you move this one line up so it isn't in the try block? `completeHeaders` doesn't throw `IllegalStateException`, and the handling of that `IllegalStateException` is for a _very_ specfic race.